src/oracle/monitor.js

```diff
@@ -0,0 +1,429 @@
+ * @param {*} oracle oracle address
+ * @param {*} feed feed name
+ * @param {*} value new feed value
+ * @param {*} id submission id which is a timestamp
+ * @param {*} actualPrice feed actual aggregated price
+ * @param {*} lastRound latest round id for which there was a submission
+ */
+const updateMetrics = (oracleName, oracle, feed, value, id, actualPrice, lastRound) => {
+    //calculate price deviation from actual value
+    let priceDeviation = Math.abs((value - actualPrice) / actualPrice) * 100
+
+    oracleSubmission.labels(oracleName, oracle, feed).set(value)
+    oracleObservation.labels(oracleName, oracle, feed).set(id)
+    oracleLastRound.labels(oracleName, oracle, feed).set(lastRound)
+    oracleDeviation.labels(oracleName, oracle, feed).set(priceDeviation)
+    actualPriceGauge.labels(feed).set(actualPrice)
+}
+
+/**
+ * Function to update balance metrics
+ * @param {*} oracleName oracle name
+ * @param {*} oracle oracle address
+ * @param {*} brand brand
+ * @param {*} value balance value to set
+ */
+const updateBalanceMetrics = (oracleName, oracle, brand, value) => {
+    oracleBalance.labels(oracleName, oracle, brand).set(value)
+}
+
+/**
+ * Function to query price for feed
+ * @param {*} feed feed like 'BRAND_IN-BRAND_OUT'
+ * @returns the price of the feed
+ */
+const queryPrice = async (feed) => {
+    try {
+        const capDataStr = await vstorage.readLatest(
+            `published.priceFeed.${feed}_price_feed`,
+        );
+
+        //parse the value
+        var capData = JSON.parse(JSON.parse(capDataStr).value)
+        capData = JSON.parse(capData.values[0])
+        //replace any extra characters
+        capData = JSON.parse(capData.body.replaceAll("\\", ""))
+
+        //get the latest price by dividing amountOut by amountIn
+        var latestPrice = Number(capData.amountOut.value.digits) / Number(capData.amountIn.value.digits)
+        amountsIn[feed] = Number(capData.amountIn.value.digits)
+
+        console.log(feed + " Price Query: " + String(latestPrice))
+        actualPriceGauge.labels(feed).set(latestPrice)
+        return latestPrice
+
+    }
+    catch (err) {
+        console.log("Price could not be obtained")
+        return 0
+    }
+
+}
+
+/**
+  * Function to get offers and balances
+  * @param {*} follower offers and balances
+  * @param {*} oracle oracle address
+  * @returns an object containing the offers and balances
+  */
+const getOffersAndBalances = async (follower, oracle) => {
+
+    let toReturn = {
+        offers: [],
+        balances: []
+    };
+    let counter = 0;
+    let lastVisited = 0;
+
+    for await (const followerElement of iterateReverse(follower)) {
+
+        if (counter == 10){
+          break;
+        }
+    
+        //if it is an offer status 
+        if (followerElement.value.updated == "offerStatus"){
+          //get id
+          let id = followerElement.value.status.id
+    
+          //if a new and final state
+          if (id != lastVisited) {
```

Comment:
```suggestion
          if (id !== lastVisited) {
```

Reply:
Fixed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/ad4d3cb5cfb0b8e6af0656b9b7e94bf94fdf671c